app/src/main/java/org/mozilla/fenix/tabtray/TabTrayView.kt

```diff
@@ -345,6 +350,54 @@ class TabTrayView(
     var mode: Mode = Mode.Normal
         private set
 
+    fun updateTabsTrayLayout() {
+        if (container.context.settings().gridTabView) {
+            setupGridTabView()
+        } else {
+            setupListTabView()
+        }
+    }
+
+    private fun setupGridTabView() {
+        view.tabsTray.apply {
+            val gridLayoutManager =
+                GridLayoutManager(container.context, getNumberOfGridColumns(container.context))
+
+            gridLayoutManager.spanSizeLookup = object : GridLayoutManager.SpanSizeLookup() {
+                override fun getSpanSize(position: Int): Int {
+                    val numTabs = tabsAdapter.itemCount
+                    return if (position < numTabs) {
+                        1
+                    } else {
+                        getNumberOfGridColumns(container.context)
+                    }
+                }
+            }
+
+            layoutManager = gridLayoutManager
+            addItemDecoration(TabTrayGridItemDecorator(container.context))
+        }
+    }
+
+    /**
+     * Returns the number of columns that will fit in the grid layout for the current screen.
+     */
+    private fun getNumberOfGridColumns(context: Context): Int {
```

Comment:
Do we also use the same logic for Top Sites? If yes - could we move this to an ext somehow?

Reply:
We use a constant 4 columns for top sites, but it would definitely be nice to move this out to some ext/util function.